### ISSUE TITLE
Feature request: post formatting step/hook

### ISSUE THREAD

**Issue by kaelig** [amzn/style-dictionary#793]
## Use case

My tokens are generated using the programmatic API, like so:

```js
// design-tokens.js
// Run `node ./design-tokens.js` to generate
import StyleDictionary from 'style-dictionary';
import config from './sd.config.js';

StyleDictionary.extend(config).buildAllPlatforms();
```

With this config:

```js
// sd.config.js
export default {
  source: ['design-system/**/*.tokens.json'],
  platforms: {
    css: {
      transformGroup: 'css',
      buildPath: 'design-system/styles/',
      files: [
        {
          destination: 'design-tokens.css',
          format: 'css/variables',
        },
      ],
    },
};
```

I need to format the output using Prettier and/or ESLint.

### npm script

Using a `postdesign-tokens` npm script runs Prettier right after `npm run design-tokens`.

```json5
// package.json
"scripts: {
  "design-tokens": "node design-tokens.js",
  "postdesign-tokens": "prettier ..."
}
```

This doesn't work for me, but might be good for most people. 

The reason it doesn't work for me is I need to generate tokens outside of npm scripts, both as part of my development server and in a browser-based environment. That's why I picked the programmatic approach to running Style Dictionary.

### My current solution

For every single existing format, we define a new custom format. This can get out of hand quickly.

```js
import formats from 'style-dictionary/lib/common/formats.js';
import prettier from 'prettier';
import process from 'process';

const prettierConfig = await prettier.resolveConfig(process.cwd());

export default {
  format: {
    // 1. Define new custom format, based on existing format
    'prettier/css/variables': (opts) => {
      // 2. Format tokens using existing CSS Variables format
      const stringifiedTokens = formats['css/variables'](opts);
      // 3. Run prettier on formatted tokens
      return prettier.format(stringifiedTokens, {
        ...prettierConfig,
        parser: 'css',
      });
    },
  },

  source: ['design-system/**/*.tokens.json'],
  platforms: {
    css: {
      transformGroup: 'css',
      buildPath: 'design-system/styles/',
      files: [
        {
          destination: 'design-tokens.css',
          // 4. Use new custom format
          format: 'prettier/css/variables',
        },
      ],
    },
  },
};
```

### Proposed solution: `postformat`

Add a new step to allow postprocessing formatted tokens.

```js
import prettier from 'prettier';
import process from 'process';

const prettierConfig = await prettier.resolveConfig(process.cwd());

export default {
  postformat: [
    (stringifiedTokens, format, opts) => {
      return prettier.format(stringifiedTokens, {
        ...prettierConfig,
        // Prettier infers the parser from the file extension
        filePath: format.destination,
      });
    },
   // Additional postformatting steps, such as PostCSS, ESLint...
  ],

  source: ['design-system/**/*.tokens.json'],
  platforms: {
    css: {
      transformGroup: 'css',
      buildPath: 'design-system/styles/',
      files: [
        {
          destination: 'design-tokens.css',
          format: 'css/variables',
        },
      ],
    },
  },
};
```

**Comment by kaelig** [comment]
>I wonder if the 1st party answer here is to just define your own formats which include the Prettier formatting. Am I missing a step?

For now, I see two solutions, see "Existing workarounds" in my original post for more details.

tl;dr:

1. npm script: `"postdesign-tokens": "prettier <options>"` that automatically runs after "npm run design-tokens"
2. defining your own formats: re-use built-in format, then run prettier formatting programmatically on the generated output

**Comment by chazzmoney** [comment]
Hi @kaelig and thanks for your question

A couple potential options that might help make your life easier:

The first is the somewhat unknown [Actions functionality](https://amzn.github.io/style-dictionary/#/actions). This is something that runs after the file has been generated by SD. If writing the file out completely, then having an action read it in a put it through prettier and write it back out again works, then we have a solution that works for you right now. If, however, you need to process the string directly (you mentioned "in browser"), then this might not work for you.

The second is just two slight modifications to your existing current solution. The first is that you should be able to use a monkey patch rather than creating a new format and the second is that you could do it in a loop for each prettier config. (I am not familiar with prettier but assume you can load multiple configurations from multiple files). Something like:

```
import styleDictionary from 'style-dictionary';
import prettier from 'prettier';

const prettierMonkeyPatches = {
  './prettier.css.config': ['css/variables'],
  './prettier.sass.config': ['scss/map-flat', 'scss/variables'],  // only patch these twos css formats
  './prettier.js.config': Object.keys(styleDictionary.formats).filter((format) => format.includes('javascript/')),  // patch all js formats
}

const styleDictionaryFormats = Object.assign({}, styleDictionary.formats)
for (const [prettierConfigFile, formatsToPrettify] of Object.entries(prettierMonkeyPatches)) {
  formatsToPrettify.forEach((format) => {
    const prettierConfig = prettier.loadConfig(prettierConfigFile)  // I'm making this prettier function up, but I assume you can do something like this
    styleDictionary.formats[format] = (opts) => {
      const stringifiedTokens = styleDictionaryFormats[format](opts)
      return prettier.format(stringifiedTokens, {
        ...prettierConfig,
        // Prettier infers the parser from the file extension
        filepath: opts.file.destination
      });
    }
  })
}

export default {
  source: ['design-system/**/*.tokens.json'],
  platforms: {
    ... // normal stuff here
  },
};
```

Note that if you have a prettier file that applies to all formats that have a substring match, you could autogenerate the prettierMonkeyPatches variable as well with a loop like this:

```
const formatsToPrettify = ['javascript', 'css', 'sass', 'scss']
let prettierMonkeyPatches = {}
formatsToPrettify.forEach((format) => {
  prettierMonkeyPatches[`prettier.${format}.config`] =  Object.keys(styleDictionary.formats).filter((format) => format.includes(`${format}/`))
```

It may be that neither of these meets your needs. However, I hope that they are helpful to you. We will consider your postFormat suggestion separately - we like the idea... it would take a bit of work to get done, and the two of us both have a lot on our plates right now. We definitely want to make sure we consider the full range of pros/cons/options around it before we jump to implementation.
